File: packages/contracts/src/fleet.ts
export type ClusterOnboardingState =
  | "pending"
  | "provisioning"
  | "ready"
  | "degraded"
  | "offboarding"
  | "offboarded";

export type ClusterBindingStatus =
  | "missing"
  | "valid"
  | "expiring"
  | "expired"
  | "revoked";

export type ObserverHealthState =
  | "unknown"
  | "healthy"
  | "degraded"
  | "unhealthy";

export interface ClusterRegistryEntry {
  id: string;
  display_name: string;
  api_endpoint: string;
  fleet_identifier: string | null;
  onboarding_state: ClusterOnboardingState;
  offboarding_state: string | null;
  created_at: string;
  updated_at: string;
}

export interface ClusterObserverBinding {
  id: string;
  cluster_id: string;
  auth_method: string;
  health_state: ObserverHealthState;
  last_observation_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface ClusterIdentityBinding {
  id: string;
  principal_id: string;
  cluster_id: string;
  cluster_username: string | null;
  cluster_groups: string[];
  binding_method: string;
  status: ClusterBindingStatus;
  expires_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface ClusterDetail extends ClusterRegistryEntry {
  observer_health: ObserverHealthState;
  last_observation_at: string | null;
}

export interface ClusterNode {
  name: string;
  status: string;
  roles: string[];
  kubelet_version: string | null;
  capacity: Record<string, string>;
  allocatable: Record<string, string>;
  taints: Array<{ key: string; effect: string; value?: string }>;
  created_at: string;
}

export interface ClusterNamespace {
  name: string;
  status: string;
  created_at: string;
}

export interface K8sEvent {
  reason: string;
  message: string;
  type: string;
  involved_object: { kind: string; name: string; namespace: string };
  last_timestamp: string | null;
  count: number;
}

File: apps/api/src/pinky_api/models/fleet.py
import uuid
from datetime import datetime

from sqlalchemy import ForeignKey, LargeBinary, String, UniqueConstraint
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column

from pinky_api.models.base import Base, TimestampMixin, gen_uuid


class ClusterRegistry(Base, TimestampMixin):
    __tablename__ = "cluster_registry"

    id: Mapped[uuid.UUID] = mapped_column(UUID, primary_key=True, default=gen_uuid)
    display_name: Mapped[str] = mapped_column(String, nullable=False)
    api_endpoint: Mapped[str] = mapped_column(String, nullable=False)
    fleet_identifier: Mapped[str | None] = mapped_column(String)
    onboarding_state: Mapped[str] = mapped_column(String, server_default="pending")
    offboarding_state: Mapped[str | None] = mapped_column(String)


class ClusterObserverBinding(Base, TimestampMixin):
    __tablename__ = "cluster_observer_bindings"

    id: Mapped[uuid.UUID] = mapped_column(UUID, primary_key=True, default=gen_uuid)
    cluster_id: Mapped[uuid.UUID] = mapped_column(UUID, ForeignKey("cluster_registry.id"), nullable=False)
    auth_method: Mapped[str] = mapped_column(String, nullable=False)
    health_state: Mapped[str] = mapped_column(String, server_default="unknown")
    last_observation_at: Mapped[datetime | None] = mapped_column()
    encrypted_credential: Mapped[bytes | None] = mapped_column(LargeBinary)
    rbac_scope: Mapped[dict] = mapped_column(JSONB, server_default="[]")


class ClusterIdentityBinding(Base, TimestampMixin):
    __tablename__ = "cluster_identity_bindings"
    __table_args__ = (UniqueConstraint("principal_id", "cluster_id"),)

    id: Mapped[uuid.UUID] = mapped_column(UUID, primary_key=True, default=gen_uuid)
    principal_id: Mapped[uuid.UUID] = mapped_column(UUID, ForeignKey("principals.id"), nullable=False)
    cluster_id: Mapped[uuid.UUID] = mapped_column(UUID, ForeignKey("cluster_registry.id"), nullable=False)
    cluster_username: Mapped[str | None] = mapped_column(String)
    cluster_groups: Mapped[list] = mapped_column(JSONB, server_default="[]")
    binding_method: Mapped[str] = mapped_column(String, nullable=False)
    status: Mapped[str] = mapped_column(String, server_default="missing")
    encrypted_token: Mapped[bytes | None] = mapped_column(LargeBinary)
    expires_at: Mapped[datetime | None] = mapped_column()
